crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1377 @@
+// SPDX-FileCopyrightText: Red Hat, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementation of SCMI and some of its protocols.
+//!
+//! This module implements SCMI infrastructure and some of the SCMI protocols.
+//! See [make_handlers] how to add support for another SCMI protocol or to add
+//! more functionality to an already implemented SCMI protocol.
+//!
+//! If you want to add new devices (e.g. SCMI bindings to some kinds of host
+//! devices), see [crate::devices] modules.
+
+use std::{
+    cmp::min,
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use itertools::Itertools;
+use log::{debug, error, info, warn};
+use thiserror::Error as ThisError;
+
+pub(crate) type MessageHeader = u32;
+
+pub(crate) const MAX_SIMPLE_STRING_LENGTH: usize = 16; // incl. NULL terminator
+
+/// Wrapper around SCMI values of the basic types SCMI defines.
+///
+/// Everything communicating to/from SCMI must be composed of them.
+// SCMI specification talks about Le32 parameter and return values.
+// VirtIO SCMI specification talks about u8 SCMI values.
+// Let's stick with SCMI specification for implementation simplicity.
+#[derive(Clone, Debug, PartialEq)]
```

Comment:
Pretty sure clippy would complain because of the lack of Eq trait implementation?

Reply:
Resolved by Manos's clippy patch.